fem/integ/bilininteg_elasticity_kernels.hpp

```diff
@@ -0,0 +1,398 @@
+                          const DofToQuad &maps, Vector &emat);
+
+/// @brief Elasticity kernel for AssembleDiagonalPA. Whole system only.
+///
+/// @param[in] dim 2 or 3
+/// @param[in] nDofs Number of scalar dofs per element.
+/// @param[in] fespace Vector (IBlock, JBlock<0) or scalar FE space.
+/// @param[in] lambda Quadrature function for first Lame param.
+/// @param[in] mu Quadrature function for first Lame param.
+/// @param[in] geom Geometric factors corresponding to fespace.
+/// @param[in] maps DofToQuad maps for one element (assume elements all same).
+/// @param QVec Scratch Q-Vector. nQuad x dim x dim x dim x dim x numEls.
+/// @param[out] diag diagonal of A. nDofs x dim x numEls.
+void ElasticityAssembleDiagonalPA(const int dim, const int nDofs,
+                                  const FiniteElementSpace &fespace, const QuadratureFunction &lambda,
+                                  const QuadratureFunction &mu, const GeometricFactors &geom,
+                                  const DofToQuad &maps, QuadratureFunction &QVec, Vector &diag);
+
+/// Templated implementation of ElasticityAddMultPA.
+template<int dim, int IBlock = -1, int JBlock = -1>
+void ElasticityAddMultPA(const int nDofs, const FiniteElementSpace &fespace,
+                         const QuadratureFunction &lambda, const QuadratureFunction &mu,
+                         const GeometricFactors &geom, const DofToQuad &maps, const Vector &x,
+                         QuadratureFunction &QVec, Vector &y)
+{
+   static_assert((IBlock < 0) == (JBlock < 0),
+                 "IBlock and JBlock must both be non-negative or strictly negative.");
+   static constexpr int d = dim;
+   static constexpr int qLower = (IBlock < 0) ? 0 : IBlock;
+   static constexpr int qUpper = (IBlock < 0) ? d : IBlock+1;
+   static constexpr int qSize = qUpper-qLower;
+   static constexpr int aLower = (JBlock < 0) ? 0 : JBlock;
+   static constexpr int aUpper = (JBlock < 0) ? d : JBlock+1;
+   static constexpr int aSize = aUpper-aLower;
+   static constexpr bool isComponent = (IBlock >= 0);
+
+   //Assuming all elements are the same
+   const auto &ir = lambda.GetIntRule(0);
+   const QuadratureInterpolator *E_To_Q_Map = fespace.GetQuadratureInterpolator(
+                                                 ir);
+   E_To_Q_Map->SetOutputLayout(QVectorLayout::byNODES);
+   //interpolate physical derivatives to quadrature points.
+   Vector junk;
+   E_To_Q_Map->Mult(x,QuadratureInterpolator::PHYSICAL_DERIVATIVES, junk,
+                    QVec, junk);
+
+   int numPoints = ir.GetNPoints();
+   int numEls = lambda.Size()/numPoints;
+   const auto lamDev = Reshape(lambda.Read(), numPoints, numEls);
+   const auto muDev = Reshape(mu.Read(), numPoints, numEls);
+   const auto J = Reshape(geom.J.Read(), numPoints, d, d, numEls);
+   auto Q = Reshape(QVec.ReadWrite(), numPoints, d, qSize, numEls);
+   const double *ipWeights = ir.GetWeights().Read();
+   MFEM_FORALL_2D(e, numEls, numPoints,1,1,
```

Comment:
This kernel (and the others in this PR) should use the newer mfem::forall functions rather than the old style macros.

Reply:
I changed all the kernels to use the new functions.